crates/stages/src/db.rs

```diff
@@ -118,77 +118,61 @@ where
         Ok((number, self.get_block_hash(number)?).into())
     }
 
-    /// Query [tables::CumulativeTxCount] table for total transaction
-    /// count block by [BlockNumHash] key
-    pub(crate) fn get_tx_count(&self, key: BlockNumHash) -> Result<NumTransactions, StageError> {
-        let count = self.get::<tables::CumulativeTxCount>(key)?.ok_or(
-            DatabaseIntegrityError::CumulativeTxCount { number: key.number(), hash: key.hash() },
-        )?;
-        Ok(count)
+    /// Query the block body by [BlockNumHash] key
+    pub(crate) fn get_block_body(&self, key: BlockNumHash) -> Result<StoredBlockBody, StageError> {
+        let body = self
+            .get::<tables::BlockBodies>(key)?
+            .ok_or(DatabaseIntegrityError::BlockBody { number: key.number() })?;
+        Ok(body)
     }
 
-    /// Get id for the first **potential** transaction in a block by looking up
-    /// the cumulative transaction count at the previous block.
-    ///
-    /// This function does not care whether the block is empty.
-    pub(crate) fn get_first_tx_id(&self, block: BlockNumber) -> Result<TxNumber, StageError> {
-        // Handle genesis block
-        if block == 0 {
-            return Ok(0)
-        }
-
-        let prev_key = self.get_block_numhash(block - 1)?;
-        self.get_tx_count(prev_key)
+    /// Query the block body by number
+    pub(crate) fn get_block_body_by_num(
+        &self,
+        number: BlockNumber,
+    ) -> Result<StoredBlockBody, StageError> {
+        let key = self.get_block_numhash(number)?;
+        self.get_block_body(key)
     }
 
-    /// Get id of the last transaction in the block.
-    /// Returns [None] if the block is empty.
-    ///
-    /// The blocks must exist in the database.
-    #[allow(dead_code)]
-    pub(crate) fn get_last_tx_id(
+    /// Query the last transition of the block by [BlockNumHash] key
+    pub(crate) fn get_last_block_transition(
```

Comment:
```suggestion
    pub(crate) fn get_block_transition(
```

We can encapsulate the "last" definition in the transition table definition